File: frameworks/dotnet/minimal-apis/Implementation/Routes/TemplateRoutes.cs
using Implementation.Components;
using Microsoft.AspNetCore.Http.HttpResults;

namespace Implementation.Routes;

/// <summary>
/// template: minimal APIs have no view layer of their own, and what ASP.NET Core ships for
/// server-side HTML is Razor. RazorComponentResult renders a component statically from an
/// endpoint, and the Web SDK compiles the component into the assembly at build.
/// </summary>
public static class TemplateRoutes
{
    public static void Map(IEndpointRouteBuilder app)
    {
        app.MapGet("/template/small", (Payloads p) => Page(p.Small));

        app.MapGet("/template/medium", (Payloads p) => Page(p.Medium));
    }

    // rb:wiring template.*
    // A native build compiles the routes with the request delegate generator, which cannot see
    // ItemsPage, because the Razor generator writes it. So the handlers return the base class.
    private static RazorComponentResult Page(Payload payload) =>
        new RazorComponentResult<ItemsPage>(new Dictionary<string, object?> { [nameof(ItemsPage.Body)] = payload });
}

File: frameworks/java/quarkus/Implementation/src/main/java/implementation/routes/TemplateRoutes.java
package implementation.routes;

import implementation.Payload;
import implementation.Payloads;
import io.quarkus.qute.CheckedTemplate;
import io.quarkus.qute.TemplateInstance;
import jakarta.ws.rs.GET;
import jakarta.ws.rs.Path;
import jakarta.ws.rs.Produces;
import jakarta.ws.rs.core.MediaType;

/**
 * template: Qute, Quarkus's own template engine, through quarkus-rest-qute. A handler returns a
 * TemplateInstance, and Quarkus REST renders it after the handler returns. The template is
 * type-safe: Qute checks every expression in it against Payload while the application is built.
 */
@Path("/template")
public class TemplateRoutes {

    // rb:wiring template.*
    @CheckedTemplate
    public static class Templates {

        /** templates/TemplateRoutes/page.html, with the payload as body. */
        public static native TemplateInstance page(Payload body);
    }

    private final Payloads p;

    TemplateRoutes(Payloads p) {
        this.p = p;
    }

    // rb:handler template.small
    @GET
    @Path("small")
    @Produces(MediaType.TEXT_HTML)
    public TemplateInstance small() {
        return Templates.page(p.small());
    }

    // rb:handler template.medium
    @GET
    @Path("medium")
    @Produces(MediaType.TEXT_HTML)
    public TemplateInstance medium() {
        return Templates.page(p.medium());
    }
}
